azkaban-common/src/main/java/azkaban/version/JdbcVersionSetLoader.java

```diff
@@ -0,0 +1,237 @@
+import azkaban.db.SQLTransaction;
+import java.io.IOException;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.commons.dbutils.ResultSetHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A JDBC based implementation for loading VersionSet from the backed db. This class keeps the copy
+ * of entries in the version_set table locally as well to reduce the number of queries to the db.
+ * <p>
+ * This is a singleton class to be instantiated using dependency injection.
+ */
+@Singleton
+public class JdbcVersionSetLoader implements VersionSetLoader {
+
+  private static final Logger logger = LoggerFactory.getLogger(JdbcVersionSetLoader.class);
+
+  private static final String SELECT_VSET_FROM_ID = "SELECT vset.id, vset.md5, vset.json FROM "
+      + "version_set vset WHERE vset.id=?";
+  private static final String SELECT_VSET_FROM_MD5 = "SELECT vset.id, vset.md5, vset.json FROM "
+      + "version_set vset WHERE vset.md5=?";
+  private static final String SELECT_ALL_VSET = "SELECT vset.id, vset.md5, vset.json FROM "
+      + "version_set vset";
+  private static final String INSERT_VSET = "INSERT INTO version_set (md5, json) values (?,?)";
+  private static final String DELETE_VSET = "DELETE FROM version_set vset WHERE vset.md5=?";
+
+  private final DatabaseOperator dbOperator;
+  private final Map<String, VersionSet> md5ToVersionSet = new HashMap<>();
+  private final Map<Integer, VersionSet> idToVersionSet = new HashMap<>();
+
+  /**
+   * This constructor should be instantiated using dependency injection only to make sure that that
+   * class remains Singleton.
+   *
+   * @param dbOperator Interface to perform db related operations.
+   */
+  @Inject
+  public JdbcVersionSetLoader(DatabaseOperator dbOperator) throws IOException {
+    this.dbOperator = dbOperator;
+    for (VersionSet versionSet : fetchAllVersionSets()) {
+      this.md5ToVersionSet.put(versionSet.getVersionSetMd5Hex(), versionSet);
+      this.idToVersionSet.put(versionSet.getVersionSetId(), versionSet);
+    }
+  }
+
+  /**
+   * This method inserts the versionSetJsonString and versionSetMd5Hex to the Table version_set and
+   * get the versionSetId autogenerated post insertion.
+   */
+  @Override
+  public synchronized int insertAndGetVersionSetId(String versionSetMd5Hex,
+      String versionSetJsonString) throws IOException {
+    final SQLTransaction<Integer> insertVersionSet = transOperator -> {
+      final VersionSetHandler versionSetHandler = new VersionSetHandler();
+      // Check if the versionSet already exists for versionSetMd5Hex
+      List<VersionSet> versionSets = transOperator
+          .query(SELECT_VSET_FROM_MD5, versionSetHandler, versionSetMd5Hex);
+      if (versionSets.size() > 1) {
+        throw new SQLException(
```

Comment:
Application specific custom exception can be thrown here. From the old code ExecutorManagerException is thrown from DAO. In this case as this will be part of imagemgmt package, ImageMgmtDaoException can be thrown here.

Reply:
Since VersionSetLoader interface has the definition to throw IOException, it is kept along the same lines.